modules/host/kernel/buildKernel.nix

```diff
@@ -0,0 +1,110 @@
+# Copyright 2022-2024 TII (SSRC) and the Ghaf contributors
+# SPDX-License-Identifier: CC-BY-SA-4.0
+{
+  config,
+  pkgs,
+  lib,
+}: {
+  src,
+  modDirVersion,
+  version,
+  kernelPatches ? [],
+}: let
+  base_kernel =
+    pkgs.linuxManualConfig rec
+    {
+      inherit src modDirVersion version kernelPatches;
+      allowImportFromDerivation = true;
+      /*
+      baseline "make tinyconfig"
+      - enabled for 64-bit, TTY, printk and initrd
+      - fixed following NixOS required assertions via "make menuconfig" + search
+      (following is documented here to highlight NixOS required (asserted) kernel features)
+      ❯ nix build .#packages.x86_64-linux.lenovo-x1-carbon-gen11-debug --accept-flake-config
+      error:
+         Failed assertions:
+         - CONFIG_DEVTMPFS is not enabled!
+         - CONFIG_CGROUPS is not enabled!
+         - CONFIG_INOTIFY_USER is not enabled!
+         - CONFIG_SIGNALFD is not enabled!
+         - CONFIG_TIMERFD is not enabled!
+         - CONFIG_EPOLL is not enabled!
+         - CONFIG_NET is not enabled!
+         - CONFIG_SYSFS is not enabled!
+         - CONFIG_PROC_FS is not enabled!
+         - CONFIG_FHANDLE is not enabled!
+         - CONFIG_CRYPTO_USER_API_HASH is not enabled!
+         - CONFIG_CRYPTO_HMAC is not enabled!
+         - CONFIG_CRYPTO_SHA256 is not enabled!
+         - CONFIG_DMIID is not enabled!
+         - CONFIG_AUTOFS4_FS is not enabled!
+         - CONFIG_TMPFS_POSIX_ACL is not enabled!
+         - CONFIG_TMPFS_XATTR is not enabled!
+         - CONFIG_SECCOMP is not enabled!
+         - CONFIG_TMPFS is not yes!
+         - CONFIG_BLK_DEV_INITRD is not yes!
+         - CONFIG_EFI_STUB is not yes!
+         - CONFIG_MODULES is not yes!
+         - CONFIG_BINFMT_ELF is not yes!
+         - CONFIG_UNIX is not enabled!
+         - CONFIG_INOTIFY_USER is not yes!
+         - CONFIG_NET is not yes!
+      ...
+      additional NixOS dependencies (fixed):
+      > modprobe: FATAL: Module uas not found ...
+      > modprobe: FATAL: Module nvme not found ...
+      ... < many packages enabled as M,
+            others allowMissing = true with overlay
+            - see implementation below under cfg.enable
+      - also see https://github.com/NixOS/nixpkgs/issues/109280
+        for the context >
+      */
+      configfile = ./ghaf_host_hardened_baseline;
```

Comment:
It is not guaranteed that this config file will work for newer/older kernels.   Is it possible to use defConfig as tinyconfig and enable other features (64-bit support, TTY, printk, and initrd etc) in extra config? It will dynamically generate minimal config for the kernel version in configuration phase.

Reply:
You are absolutely right. Config will likely break with newer and older kernels when we update nixpkgs. But our approach here is to take different way from NixOS nixpkgs kernel config builder style which accepts the new kernel config option defaults automatically. Our approach is to check if we can safely disable the new or outdated options, then disable them to keep kernel better hardened with updates.

Reply:
Thanks Vilvo for the insight. So, for Ghaf customers it is not recommended to execute 'nix flake update'. It may pull newer Linux kernel which may not work.

Reply:
In such scenario, the Ghaf-based project would be recommended https://tiiuae.github.io/ghaf/ref_impl/ghaf-based-project.html and Ghaf maintainers can choose secure defaults for new kernel config settings when newer kernels are updated.